File: SocialBookAppInfrastructure/BookService.cs
using Microsoft.EntityFrameworkCore;
using SocialBookAppApplication;
using SocialBookAppDomain;

namespace SocialBookAppInfrastructure;

public class BookService : IBookService
{
    private readonly SocialBookAppContext _context;
    private readonly IRatingService _ratingService;

    public BookService(SocialBookAppContext context, IRatingService ratingService)
    {
        _context = context;
        _ratingService = ratingService;
    }

    public double GetRatingPercentage(List<Review> reviews)
    {
        return _ratingService.GetRatingPercentage(reviews);
    }

    public List<Book> GetAllBooks()
    {
        return _context.Books
            .Include(b => b.Reviews)
            .ToList();
    }

    public Book GetBookById(int id)
    {
        var book = _context.Books
            .Include(b => b.Reviews)
            .FirstOrDefault(b => b.Id == id);
        if (book == null) throw new ArgumentException("Book not found.");
        return book;
    }

    public void AddBook(Book book)
    {
        _context.Books.Add(book);
        _context.SaveChanges();
    }

    public void DeleteBook(int id)
    {
        var book = _context.Books
            .Include(b => b.Reviews)
            .FirstOrDefault(b => b.Id == id);
        if (book == null) throw new ArgumentException("Book not found.");

        _context.Reviews.RemoveRange(book.Reviews);
        _context.Books.Remove(book);
        _context.SaveChanges();
    }

    public List<Book> GetUserBooks(int userId)
    {
        var user = _context.Users.FirstOrDefault(u => u.Id == userId);
        if (user == null) throw new ArgumentException("User not found.");
        return _context.Books.Where(b => b.Reviews.Any(r => r.User.Id == userId)).ToList();
    }

    public void AddReview(int bookId, int userId, Review review)
    {
        var book = GetBookById(bookId);
        if (book == null) throw new ArgumentException("Book not found.");
        var user = _context.Users.FirstOrDefault(u => u.Id == userId);
        if (user == null) throw new ArgumentException("User not found.");
        review.User = user;
        review.Book = book;
        review.CreatedDate = DateTime.Now;
        _context.Reviews.Add(review);
        _context.SaveChanges();
    }

    public void UpdateBook(int id, Book book)
    {
        var existing = _context.Books.FirstOrDefault(b => b.Id == id);
        if (existing == null) throw new ArgumentException("Book not found.");
        existing.Title = book.Title;
        existing.Blurb = book.Blurb;
        existing.Language = book.Language;
        existing.Genre = book.Genre;
        existing.Format = book.Format;
        existing.IsChildFriendly = book.IsChildFriendly;
        existing.Pages = book.Pages;
        existing.Chapters = book.Chapters;
        existing.Published = book.Published;
        _context.SaveChanges();
    }
}

File: SocialBookAppVue/src/services/bookService.ts
import { api } from './api'
import type { Book } from '../types/book.ts'
import type { Review } from '../types/review.ts'

export const getBooks = async () => {
  const response = await api.get('/book')
  return response.data
}

export const getBookById = async (id: number) => {
  const response = await api.get(`/book/${id}`)
  return response.data
}

export const getUserBooks = async (userId: number) => {
  const response = await api.get(`/book/user/${userId}`)
  return response.data
}

export const getRatingPercentage = async (id: number) => {
  const response = await api.get(`/book/${id}/rating`)
  return response.data
}

export const addReview = async (bookId: number, userId: number, review: Review) => {
  const response = await api.post(`/book/${bookId}/review/${userId}`, review)
  return response.data
}

export const addBook = async (book: Book) => {
  const response = await api.post('/book', book)
  return response.data
}

export const deleteBook = async (id: number) => {
  await api.delete(`/book/${id}`)
}

export const updateBook = async (id: number, book: Book) => {
  const response = await api.put(`/book/${id}`, book)
  return response.data
}
